### PR TITLE
Add EC2HibernateInstanceOperator and EC2RebootInstanceOperator

### LINKED ISSUE #35703: Add EC2RebootInstanceOperator and EC2HibernateInstanceOperator  to Amazon Provider
### Description

The Amazon Airflow Provider lacks operators for "Reboot Instance" and "Hibernate Instance," two states available in the AWS UI. Achieving feature parity would provide a seamless experience, aligning Airflow with AWS capabilities.

I'd like to see the EC2RebootInstanceOperator and EC2HibernateInstanceOperator added to Amazon Provider.

### Use case/motivation

This enhancement ensures users can manage EC2 instances in Airflow the same way they do in the AWS UI.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by dashton90**
Adds two new operators to Amazon provider package: `EC2HibernateInstanceOperator`, and `EC2RebootInstanceOperator`

closes: #35703

### CODE REVIEW

**Comment by dashton90** [comment]
Apologies, I pushed a very bad rebase to the [previous PR](https://github.com/apache/airflow/pull/35770) which requested review from a dozen contributors. Reopening a separate PR to keep things organized

**Comment by vincbeck** [comment]
> Apologies, I pushed a very bad rebase to the [previous PR](https://github.com/apache/airflow/pull/35770) which requested review from a dozen contributors. Reopening a separate PR to keep things organized

Not a problem. Tagging @o-nikolas here since he reviewed your previous PR as well.

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 1]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+
+    :param instance_ids: ID of the instance to be terminated.
+    :param aws_conn_id: AWS connection to use
+    :param region_name: AWS region name associated with the client.
+    :param poll_interval: Number of seconds to wait before attempting to
+        check state of instance. Only used if wait_for_completion is True. Default is 20.
+    :param max_attempts: Maximum number of attempts when checking state of instance.
+        Only used if wait_for_completion is True. Default is 20.
+    :param wait_for_completion: If True, the operator will wait for the instance to be
+        in the `terminated` state before returning.
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
        in the `running` state before returning.
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 2]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+
+class EC2RebootInstanceOperator(BaseOperator):
+    """
+    Reboot AWS EC2 instance using boto3.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:EC2RebootInstanceOperator`
+
+    :param instance_ids: ID of the instance to be terminated.
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    :param instance_ids: ID of the instance(s) to be terminated.
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 3]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+    :param region_name: AWS region name associated with the client.
+    :param poll_interval: Number of seconds to wait before attempting to
+        check state of instance. Only used if wait_for_completion is True. Default is 20.
+    :param max_attempts: Maximum number of attempts when checking state of instance.
+        Only used if wait_for_completion is True. Default is 20.
+    :param wait_for_completion: If True, the operator will wait for the instance to be
+        in the `terminated` state before returning.
+    """
+
+    template_fields: Sequence[str] = ("instance_id", "region_name")
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    template_fields: Sequence[str] = ("instance_ids", "region_name")
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 4]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+
+class EC2HibernateInstanceOperator(BaseOperator):
+    """
+    Hibernate AWS EC2 instance using boto3.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:EC2HibernateInstanceOperator`
+
+    :param instance_ids: ID of the instance to be terminated.
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    :param instance_ids: ID of the instance(s) to be terminated.
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 5]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+    :param aws_conn_id: AWS connection to use
+    :param region_name: AWS region name associated with the client.
+    :param poll_interval: Number of seconds to wait before attempting to
+        check state of instance. Only used if wait_for_completion is True. Default is 20.
+    :param max_attempts: Maximum number of attempts when checking state of instance.
+        Only used if wait_for_completion is True. Default is 20.
+    :param wait_for_completion: If True, the operator will wait for the instance to be
+        in the `terminated` state before returning.
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
        in the `stopped` state before returning.
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 6]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+    """
+
+    template_fields: Sequence[str] = ("instance_id", "region_name")
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    template_fields: Sequence[str] = ("instance_ids", "region_name")
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 7]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+                WaiterConfig={
+                    "Delay": self.poll_interval,
+                    "MaxAttempts": self.max_attempts,
+                },
+            )
+
+
+class EC2HibernateInstanceOperator(BaseOperator):
+    """
+    Hibernate AWS EC2 instance using boto3.
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    Hibernate Amazon EC2 instance.
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 8]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
                         "MaxAttempts": self.max_attempts,
                     },
                 )
+
+
+class EC2RebootInstanceOperator(BaseOperator):
+    """
+    Reboot AWS EC2 instance using boto3.
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    Reboot Amazon EC2 instance.
```

**Review by vincbeck — COMMENTED** [review]
LGTM overall, only some nits

**Review by vincbeck — APPROVED** [review]

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 9]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+
+class EC2RebootInstanceOperator(BaseOperator):
+    """
+    Reboot Amazon EC2 instance.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:EC2RebootInstanceOperator`
+
+    :param instance_ids: ID of the instance(s) to be terminated.
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    :param instance_ids: ID of the instance(s) to be rebooted.
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 10]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
                         "MaxAttempts": self.max_attempts,
                     },
                 )
+
+
+class EC2RebootInstanceOperator(BaseOperator):
+    """
+    Reboot Amazon EC2 instance.
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    Reboot Amazon EC2 instances.
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 11]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+
+class EC2HibernateInstanceOperator(BaseOperator):
+    """
+    Hibernate Amazon EC2 instance.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:EC2HibernateInstanceOperator`
+
+    :param instance_ids: ID of the instance(s) to be terminated.
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    :param instance_ids: ID of the instance(s) to be hibernated.
```

File: airflow/providers/amazon/aws/operators/ec2.py [hunk 12]
```diff
@@ -254,3 +255,128 @@ def execute(self, context: Context):
...
+                WaiterConfig={
+                    "Delay": self.poll_interval,
+                    "MaxAttempts": self.max_attempts,
+                },
+            )
+
+
+class EC2HibernateInstanceOperator(BaseOperator):
+    """
+    Hibernate Amazon EC2 instance.
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/ec2.py]:
```suggestion
    Hibernate Amazon EC2 instances.
```

File: tests/system/providers/amazon/aws/example_ec2.py [hunk 13]
```diff
@@ -142,6 +145,22 @@ def parse_response(instance_ids: list):
     )
     # [END howto_sensor_ec2_instance_state]
 
+    # [START howto_operator_ec2_reboot_instance]
+    reboot_instance = EC2RebootInstanceOperator(
+        task_id="reboot_instace",
+        instance_ids=instance_id,
+    )
+    # [END howto_operator_ec2_reboot_instance]
+    reboot_instance.trigger_rule = TriggerRule.ALL_DONE
```
Comment (o-nikolas) [inline tests/system/providers/amazon/aws/example_ec2.py]: We don't need `TriggerRule.ALL_DONE` here (or on the hibernate), that's only for steps that are needed for cleanup, to ensure they run and don't leave behind resources. So it's only really needed on the terminate instance operator.

**Review by o-nikolas — COMMENTED** [review]
Looking really good!

Just a few more things to fix

File: tests/system/providers/amazon/aws/example_ec2.py (hunk 13 above)
Comment (dashton90) [reply tests/system/providers/amazon/aws/example_ec2.py]: Removed in https://github.com/apache/airflow/pull/35790/commits/57ad259b0b90d0ad8626e24f25061bca8947ddbf

**Review by o-nikolas — APPROVED** [review]

**Comment by ferruzzi** [comment]
This merge broke the system tests;  t4g.micro instances can not be hibernated.  Did you test this out?  If so, I don't suppose you remember which instance size worked for you?

```
botocore.exceptions.ClientError: An error occurred (UnsupportedHibernationConfiguration) when calling the RunInstances operation: The instance type does not support hibernation: t4g.micro. For information about supported instance types, see the Amazon EC2 Hibernation documentation.
```

**Comment by ferruzzi** [comment]
Looking through the code, it looks like you are checking for a hibernation option, but the exception you are raising `raise AirflowException(f"Instance {instance['InstanceId']} is not configured for hibernation")` is not making it into the system test's logs, so we may need to have another look at the unit tests as well.

**Comment by dashton90** [comment]
I tested on an m4.large instance

And didn't come across as harsh at all. Please let me know how I can assist with the system tests.

**Comment by dashton90** [comment]
As a separate note, are these tests available publicly? Would be happy to follow any additional steps before opening a PR in the future

**Comment by ferruzzi** [comment]
What do you think about the unit tests?  Do they need to be tweaked?

> I tested on an m4.large instance

Cool.   I'll see if I can find documentation around which sizes DO work, but at least we have a place to start now.

> are these tests available publicly?

We (the AWS Airflow team) run most of the example_{foo} tests in the system test folder on a regular basis and report the results [here](https://aws-mwaa.github.io/#/open-source/system-tests/dashboard.html).  The logs aren't publicly available (yet??), unfortunately, but you can see a pass/fail for the last 10 runs there.

**Comment by dashton90** [comment]
> Cool. I'll see if I can find documentation around which sizes DO work, but at least we have a place to start now.

The supported instances are listed [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/hibernating-prerequisites.html#hibernation-prereqs-supported-instance-families)

> What do you think about the unit tests? Do they need to be tweaked?

If the logs are not making it then the tests should definitely be updated. I did add two unhappy path unit tests for trying to hibernate instances not configured for hibernation [1](https://github.com/apache/airflow/pull/35790/files#diff-04a8f0131e1098782c138b484b4e563bcda622b1442371be798eeef95d004a55R267) [2](https://github.com/apache/airflow/pull/35790/files#diff-04a8f0131e1098782c138b484b4e563bcda622b1442371be798eeef95d004a55R293).

> We (the AWS Airflow team) run most of the example_{foo} tests in the system test folder on a regular basis and report the results [here](https://aws-mwaa.github.io/#/open-source/system-tests/dashboard.html). The logs aren't publicly available (yet??), unfortunately, but you can see a pass/fail for the last 10 runs there.

Amazing, I'll keep an eye on this page next time I contribute :)

**Comment by ferruzzi** [comment]
> The supported instances are listed [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/hibernating-prerequisites.html#hibernation-prereqs-supported-instance-families)

You took the words out of my ~~mouth~~ fingers.  

```
$ aws ec2 describe-instance-types --filters Name=hibernation-supported,Values=true --query "InstanceTypes[*].[InstanceType]" --output text | sort | grep "micro"
t2.micro
t3a.micro
t3.micro
```  

EC2 instance types aren't my forte, but it looks like swapping the test over to t3.micro instance type should get it passing, just running on an x86 CPU vs an ARM CPU (I think).   I'll run the test on the t3.micro and see if it passes.

**Comment by dashton90** [comment]
Just re-reading through your comments:

> Looking through the code, it looks like you are checking for a hibernation option, but the exception you are raising raise AirflowException(f"Instance {instance['InstanceId']} is not configured for hibernation") is not making it into the system test's logs, so we may need to have another look at the unit tests as well.

This check is only done when you run the `EC2HibernateInstanceOperator`. The example_ec2 DAG would never get to this operator since it is failing during instance creation.

If you'd like, I can update the example to use an m4 or t3 instance.

**Comment by ferruzzi** [comment]
> If you'd like, I can update the example to use an m4 or t3 instance.

I'm running it against my live AWS account.  It is going need a few other changes, too.   The AMI architecture needs to be adjusted [here](https://github.com/apache/airflow/blob/main/tests/system/providers/amazon/aws/example_ec2.py#L57) if we change from a t4g.  When I do that it fails with "botocore.exceptions.ClientError: An error occurred (UnsupportedHibernationConfiguration) when calling the RunInstances operation: For hibernation, the root device volume must be encrypted."   We're on the right track, though.   Do you want to poke at it and get it sorted out, or do you want me to?

**Comment by ferruzzi** [comment]
> The example_ec2 DAG would never get to this operator since it is failing during instance creation.

Yup, good catch.  My mistake.

**Comment by dashton90** [comment]
> Do you want to poke at it and get it sorted out, or do you want me to?

If you're already on it then I'll leave it to you. But happy to take over if you need another set of eyes

**Comment by ferruzzi** [comment]
It looks like in order to get this working we're going to need to set up a KMS key and manually define the device mappings to configure an EBS volume.  I won't be able to get this done until Monday.  Should you feel like taking care of it before then, have at it and ping me in the PR.

[This](https://aws.amazon.com/blogs/security/how-to-quickly-launch-encrypted-ebs-backed-ec2-instances-from-unencrypted-amis/) isn't directly applicable since it isn't using the boto3 API, but it should have the info we need, just in a different format.

**Comment by dashton90** [comment]
Created https://github.com/apache/airflow/pull/35839. I've tested the changes independently, but can I leave it to you to do the end-to-end tests?

**Comment by ferruzzi** [comment]
Amazing, thanks!   I'll have a look later today.

**Pull request merged**